WordPress/WordPressTest/BloggingPromptsServiceTests.swift

```diff
@@ -136,39 +138,172 @@ final class BloggingPromptsServiceTests: CoreDataTestCase {
+        let date = try passedDate()
+        let dateComponents = Self.calendar.dateComponents(in: Self.utcTimeZone, from: date)
+        let year = try passedParameter("force_year") as? Int
+        XCTAssertEqual(year, expectedDateComponents.year)
+        XCTAssertEqual(dateComponents.month, expectedDateComponents.month)
+        XCTAssertEqual(dateComponents.day, expectedDateComponents.day)
+
+        let numberParameter = try XCTUnwrap(passedNumber())
+        XCTAssertEqual(numberParameter, expectedNumber)
+    }
+
+    // MARK: - Upsert Tests
+
+    // new prompts should overwrite any
+    func test_fetchPrompt_shouldOverwritePromptsWithExistingDates() throws {
+        // use actual remote object so the request can be intercepted by HTTPStubs.
+        service = BloggingPromptsService(contextManager: contextManager, blog: blog)
+        stubFetchPromptsResponse()
+
+        let expectedPromptIDs: Set<Int> = [239, 248]
+
+        // insert existing prompts.
+        let date = try XCTUnwrap(Self.dateFormatter.date(from: "2022-05-03"))
+        makeBloggingPrompt(siteID: Int32(siteID), date: date)
+        contextManager.save(contextManager.mainContext)
+
+        let expectation = expectation(description: "Fetch prompts should succeed")
+        service.fetchPrompts(from: .distantPast) { prompts in
+            // the existing prompt should have been overwritten.
+            XCTAssertEqual(prompts.count, 2)
+
+            let promptIDs = Set(prompts.map { Int($0.promptID) })
+            XCTAssertTrue(expectedPromptIDs.elementsEqual(promptIDs))
+
+            expectation.fulfill()
+
+        } failure: { error in
+            XCTFail("This closure shouldn't be called.")
+            expectation.fulfill()
+        }
+
+        wait(for: [expectation], timeout: timeout)
+    }
+
+    // there should only be one prompt per date.
+    func test_fetchPrompt_shouldDeleteExcessPromptsWithTheSameDates() throws {
+        // use actual remote object so the request can be intercepted by HTTPStubs.
+        service = BloggingPromptsService(contextManager: contextManager, blog: blog)
+        stubFetchPromptsResponse()
+
+        let expectedPromptIDs: Set<Int> = [239, 248]
+
+        // insert existing prompts.
+        let date = try XCTUnwrap(Self.dateFormatter.date(from: "2022-05-03"))
+        makeBloggingPrompt(siteID: Int32(siteID), date: date)
+        makeBloggingPrompt(siteID: Int32(siteID), date: date)
+        makeBloggingPrompt(siteID: Int32(siteID), date: date)
+        makeBloggingPrompt(siteID: Int32(siteID), date: date)
+        makeBloggingPrompt(siteID: Int32(siteID), date: date)
+        contextManager.save(contextManager.mainContext)
+
+        let expectation = expectation(description: "Fetch prompts should succeed")
+        service.fetchPrompts(from: .distantPast) { prompts in
+            // the existing prompt should have been overwritten.
+            XCTAssertEqual(prompts.count, 2)
```

Comment:
Out of scope because the test was already setup this way, but as a reader with no context, it took me a while to patch together that we are doing `prompts.count == 2` as a check for the existing prompts having been deleted because in the HTTP stub we set above we read from a JSON that has two prompts. 

Did you have context getting into this test? Was that clearer for you?

Reply:
> it took me a while to patch together that we are doing prompts.count == 2 as a check for the existing prompts having been deleted because in the HTTP stub we set above we read from a JSON that has two prompts. 
> Did you have context getting into this test?

Yes, I did have context for this test because I was the one who created the test originally. However, I see your point, since the assertions are hardcoded... and it would take some extra steps to connect the dots and understand that this is closely tied to the values in the stub file.

I've tried to get rid of almost all the hardcoded values (save for a few assumptions about the test data) through https://github.com/wordpress-mobile/WordPress-iOS/pull/21952/commits/176c53f8d9fe2409ec84b528bb805556f1fb2da8. Let me know if this is more readable for you overall. 🙂

Reply:
Thanks @dvdchr . Your changes are definitely an improvement. They make it clearer where the expected data comes from.

I think there's room for improvement in our HTTP stubs. Personally, I'd like to move away from JSONs, which keep the source of truth for the expected data in a different file than the test, and have something were we can set the expected values as close as possible to where we assert them. But that's a discussion for another day and a change that would be far wider in reach. 👍